docs/docs/reference-react-component.md

```diff
@@ -0,0 +1,170 @@
+
+For example, this code ensures that the `color` prop is a string:
+
+```js
+class CustomButton extends React.Component {
+  // ...
+}
+
+CustomButton.propTypes = {
+  name: React.PropTypes.string
+};
+```
+
+### `.render()`
+
+The `render()` method is required.
+
+When called, it should examine `this.props` and `this.state` and return a single React element. This element can be either a representation of a native DOM component, such as `<div />`, or another composite component that you've defined yourself.
+
+You can also return `null` or `false` to indicate that you don't want anything rendered. React will just render a comment tag. When returning `null` or `false`, `ReactDOM.findDOMNode(this)` will return `null`.
+
+The `render()` function should be pure, meaning that it does not modify component state, it returns the same result each time it's invoked, and it does not directly interact with the browser. If you need to interact with the browser, perform your work in `componentDidMount()` or the other lifecycle methods instead. Keeping `render()` pure makes components easier to think about.
+
+### `.setState(nextState, callback)`
+
+```javascript
+void setState(
+  function|object nextState,
+  [function callback]
+)
+```
+
+Performs a shallow merge of nextState into current state. This is the primary method you use to trigger UI updates from event handlers and server request callbacks.
+
+The first argument can be an object (containing zero or more keys to update) or a function (of state and props) that returns an object containing keys to update.
+
+Here is the simple object usage:
+
+```javascript
+this.setState({mykey: 'my new value'});
+```
+
+It's also possible to pass a function with the signature `function(state, props) => newState`. This can be useful in some cases when you want to enqueue an atomic update that consults the previous value of state and props before setting any values. For instance, suppose we wanted to increment a value in state:
+
+```javascript
+this.setState((prevState, currentProps) => {
+  return {myInteger: prevState.myInteger + 1};
+});
+```
+
+The second parameter is an optional callback function that will be executed once `setState` is completed and the component is re-rendered.
+
+`setState()` does not immediately mutate `this.state` but creates a pending state transition. Accessing `this.state` after calling this method can potentially return the existing value.
+
+There is no guarantee of synchronous operation of calls to `setState` and calls may be batched for performance gains.
+
+`setState()` will always trigger a re-render unless conditional rendering logic is implemented in `shouldComponentUpdate()`. If mutable objects are being used and the logic cannot be implemented in `shouldComponentUpdate()`, calling `setState()` only when the new state differs from the previous state will avoid unnecessary re-renders.
```

Comment:
> `setState()` will always trigger a re-render unless...

Could we change this slightly to say "will eventually lead to a re-render" to account for batching? Current wording could imply a 1-to-1 correspondence between `setState` and renders.